File: Python/Algorithms/Find-the-Duplicate-Number.py
# Time: (n log(n))
# Space: O(1)

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        nums.sort()
        
        visited = nums[0]
        for i in range(1, len(nums)):
            if visited == nums[i]:
                return nums[i]
            visited = nums[i]

            
#------------------------------------------------------------------------#

# Time: O(n)
# Space: O(n)

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        visited = {}
        
        for i in nums:
            if i in visited:
                visited[i]+=1
            else:
                visited[i] = 1
        
        for i in visited.keys():
            if visited[i]>1:
                return i

File: Golang/Find-the-Duplicate-Number.go
/*
  Time: O(n log(n))
  Space: O(1)
*/
func findDuplicate(nums []int) int {
    
    sort.Slice(nums, func(i, j int) bool {
        return nums[i]<nums[j]
    })
    
    visited := nums[0]
    
    for i := 1; i<len(nums); i++ {
        if nums[i] == visited {
            return nums[i]
        }
        visited = nums[i]
    }
    return -1
}

//-----------------------------------------------------------------------//

/*
   Time: O(n)
   Space: O(n)
*/

func findDuplicate(nums []int) int {
    
    visited := map[int]int{}
    
    for _, val := range nums {
        if _, exists := visited[val]; exists {
            visited[val]++
        } else {
            visited[val] = 1
        }
    }
    
    for key,_ := range visited {
        if visited[key] > 1 {
            return key
        }
    }
    return -1
}
